lib/cdo/throttle.rb

```diff
@@ -0,0 +1,45 @@
+require 'cdo/shared_cache'
+require 'dynamic_config/dcdo'
+
+module Cdo
+  module Throttle
+    CACHE_PREFIX = "cdo_throttle/".freeze
+
+    # @param [String] id - Unique identifier to throttle on.
+    # @param [Integer] limit - Number of requests allowed over period.
+    # @param [Integer] period - Period of time in seconds.
+    # @param [Integer] throttle_for - How long id should stay throttled in seconds. Optional.
+    # Defaults to Cdo::Throttle.throttle_time.
+    # @returns [Boolean] Whether or not the request should be throttled.
+    def self.throttle(id, limit, period, throttle_for = throttle_time)
+      full_key = CACHE_PREFIX + id.to_s
+      value = CDO.shared_cache.read(full_key) || empty_value
```

Comment:
Does this result in a network call?  How is failure handled (e.g. if the shared cache is down)?

Reply:
yes, this will result in a network call if `CDO.shared_cache` is using ElastiCache (it's the default, but falls back to a FileStore cache if initializing ElastiCache fails [here](https://github.com/code-dot-org/code-dot-org/blob/d79376aae018a417d3fb1882589868c8d207bc2d/lib/cdo/shared_cache.rb#L34-L40)).

if i'm reading the implementation correctly, `nil` will be returned on failure. [this is rails' implementation of `read`](https://github.com/rails/rails/blob/914caca2d31bd753f47f9168f2a375921d9e91cc/activesupport/lib/active_support/cache.rb#L346-L380), which calls the subclass' implementation of `read_entry`, which would be [this implementation when using ElastiCache](https://github.com/rails/rails/blob/914caca2d31bd753f47f9168f2a375921d9e91cc/activesupport/lib/active_support/cache/mem_cache_store.rb#L141-L144).

Reply:
I think we may want to make this more explicit and just bypass the throttler if elasticache is temporarily unavailable?